### PR TITLE
XLA CPU Fft no core framework and single thread

### PR DESCRIPTION
**Pull request by nuchi**
See https://github.com/tensorflow/tensorflow/issues/18588

Does two things:
1. Fixes the XLA CPU implementation of the fast fourier transform to avoid reliance on the tf core framework.
2. Allows AOT compilation of FFT to single-threaded environments.

See the linked issue for a longer discussion. I ran the [`randomized_tests.cc`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tests/randomized_tests.cc) suite and it passed with `--tf_xla_test_use_jit=false --tf_xla_test_device=CPU:0`. None of the tests (neither my changes nor any other ops) would run with `--tf_xla_test_use_jit=true`; I don't know if that's a problem with my setup.

Happy to make changes.

### CODE REVIEW

**Comment by nuchi** [comment]
I signed it!

File: tensorflow/compiler/xla/service/cpu/runtime_fft_impl.h [hunk 1]
```diff
@@ -179,7 +172,6 @@ template <int FFTRank, typename EigenDevice>
 void EigenFftWithRank(const EigenDevice& device, void* out, void* operand,
                       int32 fft_type, int64 input_batch, int64 fft_length0,
                       int64 fft_length1, int64 fft_length2) {
-  CHECK(::xla::FftType_IsValid(fft_type)) << fft_type;
```
Comment (sanjoy) [inline tensorflow/compiler/xla/service/cpu/runtime_fft_impl.h]: How about

```
if (!::xla::FftType_IsValid(fft_type)) {
  abort();
}
```

etc.?

**Review by sanjoy — COMMENTED** [review]
For testing, perhaps you can try testing `//tensorflow/compiler/tests:fft_test`?

File: tensorflow/compiler/xla/service/cpu/runtime_fft_impl.h (hunk 1 above)
Comment (nuchi) [reply tensorflow/compiler/xla/service/cpu/runtime_fft_impl.h]: I added the aborts in `default` at the bottom (for both rank and type). If you think it'd be cleaner to have it at the top too, let me know.

**Comment by nuchi** [comment]
Ok, I figured out that I had to manually execute `bazel-bin/tensorflow/compiler/tests/fft_test_cpu --test_device=CPU --types=DT_HALF,DT_FLOAT,DT_DOUBLE,DT_INT32,DT_INT64,DT_BOOL,DT_COMPLEX64`, and then it passes all tests.

If I just run `bazel test --config=opt //tensorflow/compiler/tests:fft_test` (which calls `--test_device=XLA_CPU`; see [build_defs.bzl](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tests/build_defs.bzl)) then it fails with the same error as the randomized tests error:
```
2018-04-21 01:14:32.248668: I tensorflow/compiler/xla/service/service.cc:159] XLA service 0x7f928643e070 executing computations on platform Host. Devices:
2018-04-21 01:14:32.248701: I tensorflow/compiler/xla/service/service.cc:167]   StreamExecutor device (0): <undefined>, <undefined>
2018-04-21 01:14:32.780468: F tensorflow/compiler/xla/service/cpu/cpu_executable.cc:69] Check failed: sym Symbol cluster_0__XlaCompiledKernel_true__XlaNumConstantArgs_0__XlaNumResourceArgs_0_.v31 not found.
Abort trap: 6
```
I'm not sure if that means that I didn't build something I was supposed to build, or if there's some subtle problem with my changes, or something else.

**Comment by ekelsen** [comment]
@sanjoy could you please continue the review after @nuchi's comments?

**Review by sanjoy — APPROVED** [review]

**Comment by nuchi** [comment]
Thanks! Out of curiosity, do you know of a simple explanation for the test failure I saw?

**Comment by sanjoy** [comment]
> Thanks! Out of curiosity, do you know of a simple explanation for the test failure I saw?

Are you able to reproduce the failure?  If so, can you please post the full log output with `--v=2 --logtostderr`?

**Comment by sanjoy** [comment]
I can't reproduce this failure with your change.  Are you possibly based on an older git revision?  The test passed with your changes merged with git sha1 `1863070e234a4a1c34a70ff73896bc9eff29fc70`.

**Comment by nuchi** [comment]
From about a week ago. If you can't reproduce then it's probably fine.

**Comment by nuchi** [comment]
Does anything else need to happen on my end? I wasn't sure what `Ubuntu CC Expected — Waiting for status to be reported` meant and whether it is supposed to take a long time.

**Comment by sanjoy** [comment]
I don't know what the Ubuntu CC bit is either.  @martinwicke ?

**Comment by martinwicke** [comment]
If you're happy with this PR, set the "awaiting testing" and "kokoro:force-run" labels. That'll trigger the tests.

**Comment by nuchi** [comment]
Does it need to keep the "await testing and then merge" label to get merged? I noticed all the tests passed.

**Pull request merged**